pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+})
+
+func testCreateFederatedResource() {
+	var (
+		resource *corev1.Pod
+		expected *unstructured.Unstructured
+		scheme   *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resource = newPod("test-pod", "nginx")
+		scheme = runtime.NewScheme()
+	})
+
+	AfterEach(func() {
+		resource = nil
+		expected = nil
+		scheme = nil
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		When("failing to convert to Unstructured", func() {
+			It("should throw an error", func() {
+				var expectedType *UnstructuredConversionError
+				_, err := createFederatedResource(scheme, resource)
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(BeAssignableToTypeOf(expectedType))
+			})
+
+			It("should return a null expected", func() {
+				expected, _ = createFederatedResource(scheme, resource)
+				Expect(expected).To(BeNil())
+			})
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			err := corev1.AddToScheme(scheme)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should not modify the input resource but copy it", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(resource).ToNot(BeIdenticalTo(expected))
+		})
+
+		It("should return no error on success", func() {
+			_, err := createFederatedResource(scheme, resource)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("the returned resource's Kind should be of Federated<kind>", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			expectedKind := federatedKindPrefix + resource.GroupVersionKind().Kind
+			Expect(expected.GroupVersionKind().Kind).
+				To(Equal(expectedKind))
+		})
+
+		It("the returned resource's Group should be the default KubeFed FederatedGroup", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Group).
+				To(Equal(kubefedopt.DefaultFederatedGroup))
+		})
+
+		It("the returned resource's Version should be the default KubeFed FederatedVersion", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Version).
+				To(Equal(kubefedopt.DefaultFederatedVersion))
+		})
+
+		It("the returned resource's name should be the same as the input's object", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GetName()).
+				To(Equal(resource.GetName()))
+		})
+
+		It("the returned resource's template should be the input's unstructured", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			expected, _ = createFederatedResource(scheme, resource)
+
+			mapObj, found, err := getTemplateField(expected)
+			targetResource := &unstructured.Unstructured{}
+
+			convErr := scheme.Convert(resource, targetResource, nil)
+			Expect(convErr).ToNot(HaveOccurred())
+			removeUnwantedFields(targetResource)
```

Comment:
One thing I try to avoid in UTs is relying on internal aspects of the production code for verification. Here you're using `removeUnwantedFields` to build the expected data which is an internal helper function also used by the code you're testing. So naturally the expected and actual will tend to match even though it might not be correct. So instead, I try to think about what the actual result should be from a blackbox perspective and setup assertions based on that. In this case, what we really expect is for none of the **metadata** fields in the original resource to be present in the federated resource template's **metadata** and thus we can simply verify the associated map is empty:

`Expect(mapObj["metadata"]).To(BeEmpty())`

Reply:
Agreed, I was thinking of the same think while writing my other tests this week.

Reply:
You are 100% right. Just be mindful that this and some of your other comments are a bit condescending, Tom.